hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyRequest.java

```diff
@@ -250,91 +249,113 @@ public EncryptedKeyVersion run() throws IOException {
    * @return OMClientResponse
    */
   @SuppressWarnings("parameternumber")
-  protected OMClientResponse prepareCreateKeyResponse(@Nonnull KeyArgs keyArgs,
+  protected OMClientResponse prepareCreateKeyResponse(
+      @Nonnull OMResponse.Builder omResponse, @Nonnull KeyArgs keyArgs,
       OmKeyInfo omKeyInfo, @Nonnull List<OmKeyLocationInfo> locations,
-      FileEncryptionInfo encryptionInfo, @Nullable IOException exception,
+      FileEncryptionInfo encryptionInfo,
       long clientID, long transactionLogIndex, @Nonnull String volumeName,
       @Nonnull String bucketName, @Nonnull String keyName,
       @Nonnull OzoneManager ozoneManager, @Nonnull OMAction omAction,
       @Nonnull PrefixManager prefixManager,
       @Nullable OmBucketInfo omBucketInfo) {
 
-    OMResponse.Builder omResponse = OMResponse.newBuilder()
-        .setStatus(OzoneManagerProtocolProtos.Status.OK);
     OMMetadataManager omMetadataManager = ozoneManager.getMetadataManager();
-
-    Map<String, String> auditMap = buildKeyArgsAuditMap(keyArgs);
-
     OMClientResponse omClientResponse = null;
-    if (exception == null) {
-      if (omKeyInfo == null) {
-        // the key does not exist, create a new object, the new blocks are the
-        // version 0
-        omKeyInfo = createKeyInfo(keyArgs, locations, keyArgs.getFactor(),
-            keyArgs.getType(), keyArgs.getDataSize(), encryptionInfo,
-            prefixManager, omBucketInfo, transactionLogIndex);
-      }
+    IOException exception = null;
 
-      long openVersion = omKeyInfo.getLatestVersionLocations().getVersion();
+    Map<String, String> auditMap = buildKeyArgsAuditMap(keyArgs);
 
-      // Append blocks
-      try {
-        omKeyInfo.appendNewBlocks(keyArgs.getKeyLocationsList().stream()
-            .map(OmKeyLocationInfo::getFromProtobuf)
-            .collect(Collectors.toList()), false);
+    if (omKeyInfo == null) {
+      // the key does not exist, create a new object, the new blocks are the
+      // version 0
+      omKeyInfo = createKeyInfo(keyArgs, locations, keyArgs.getFactor(),
+          keyArgs.getType(), keyArgs.getDataSize(), encryptionInfo,
+          prefixManager, omBucketInfo, transactionLogIndex);
+    }
 
-      } catch (IOException ex) {
-        exception = ex;
-      }
+    long openVersion = omKeyInfo.getLatestVersionLocations().getVersion();
 
-      if (exception != null) {
-        LOG.error("{} failed for Key: {} in volume/bucket:{}/{}",
-            omAction.getAction(), keyName, bucketName, volumeName, exception);
-        omClientResponse = createKeyErrorResponse(ozoneManager.getMetrics(),
-            omAction, exception, omResponse);
+    // Append blocks
```

Comment:
We are adding blocks twice for the case when keyInfo != null, as once in prepareKeyInfo and here again.
So, can we move the creation of KeyInfo logic to prepareKeyInfo and use prepareCreateKeyResponse just to generate OmClientResponse.

Reply:
I think we are adding blocks only once. But still, if we consolidate the logic it will be good. Okay with doing that in a new Jira also.